spinalcordtoolbox/scripts/sct_maths.py

```diff
@@ -270,223 +315,465 @@ def __call__(self, _, namespace, values, option_string=None):
+
+    # adjust sigma based on voxel size
+    pdims = current_value.dim[4:8]
+    sigmas = [s/d for s, d in zip(sigmas, pdims)]
+
+    return Image(
+        sct_math.smooth(current_value.data, sigmas),
+        hdr=current_value.hdr,
+    )
+
+
+def apply_laplacian(current_value, *sigmas):
+    """Implementation for -laplacian."""
+    ndims = current_value.data.ndims
+    if len(sigmas) == 1:
+        sigmas *= ndims
+    elif len(sigmas) != ndims:
+        raise SctMathsValueError(f"expected 1 or {ndims} values, got {len(sigmas)} values")
+
+    # adjust sigma based on voxel size
+    pdims = current_value.dim[4:8]
+    sigmas = [s/d for s, d in zip(sigmas, pdims)]
+
+    return Image(
+        sct_math.laplacian(current_value.data, sigmas),
+        hdr=current_value.hdr,
+    )
+
+
+def apply_denoise(current_value, patch_radius, block_radius):
+    """Implementation for -denoise."""
+    return Image(
+        sct_math.denoise_nlmeans(current_value.data, patch_radius, block_radius),
+        hdr=current_value.hdr,
+    )
+
+
+def apply_mi(current_value, fname):
+    """Implementation for -mi."""
+    return (current_value, Image(fname), 'mi', 'Mutual information')
+
+
+def apply_minorm(current_value, fname):
+    """Implementation for -minorm."""
+    return (current_value, Image(fname), 'minorm', 'Normalized Mutual information')
+
+
+def apply_corr(current_value, fname):
+    """Implementation for -corr."""
+    return (current_value, Image(fname), 'corr', 'Pearson correlation coefficient')
+
+
+def apply_symmetrize(current_value, axis):
+    """Implementation for -symmetrize."""
+    return Image(
+        sct_math.symmetrize(current_value.data, axis),
+        hdr=current_value.hdr,
+    )
+
+
+# the implementation function for each sct_maths operation
+APPLY = {
+    "add": apply_add,
+    "sub": apply_sub,
+    "mul": apply_mul,
+    "div": apply_div,
+    "mean": apply_mean,
+    "rms": apply_rms,
+    "std": apply_std,
+    "bin": apply_bin,
+    "otsu": apply_otsu,
+    "adap": apply_adap,
+    "otsu_median": apply_otsu_median,
+    "percent": apply_percent,
+    "thr": apply_thr,
+    "uthr": apply_uthr,
+    "dilate": apply_dilate,
+    "erode": apply_erode,
+    "smooth": apply_smooth,
+    "laplacian": apply_laplacian,
+    "denoise": apply_denoise,
+    "mi": apply_mi,
+    "minorm": apply_minorm,
+    "corr": apply_corr,
+    "symmetrize": apply_symmetrize,
+}
+
+
+# MAIN
+# ==========================================================================================
+def main(argv: Sequence[str]):
+    parser = get_parser()
+    arguments = parser.parse_args(argv)
+    verbose = arguments.v
+    set_loglevel(verbose=verbose)
+
+    # Check data types
+    current_type = 'number' if isinstance(arguments.i, float) else 'image'
+    for operation, args, kwargs in arguments.todo:
+        try:
+            current_type = CHECK[operation](current_type, *args, **kwargs)
```

Comment:
It turns out that simple type checking and simple program-running have basically the same structure: we just have to replace the `CHECK` dictionary with `APPLY`.